sogs/omq.py

```diff
@@ -6,62 +6,63 @@
 
 from . import crypto, config
 from .postfork import postfork
+from .model.manager import Manager
 
-omq = None
-mule_conn = None
-test_suite = False
+omq_global = None
 
+class OMQ:
+    @postfork
+    def __init__(self):
+        try:
+            import uwsgi
+        except ModuleNotFoundError:
+            return
 
-def make_omq():
-    omq = oxenmq.OxenMQ(privkey=crypto._privkey.encode(), pubkey=crypto.server_pubkey.encode())
+        self._omq = oxenmq.OxenMQ(
+            privkey=crypto._privkey.encode(), pubkey=crypto.server_pubkey.encode()
+        )
+        self._omq.ephemeral_routing_id = True
 
-    # We have multiple workers talking to the mule, so we *must* use ephemeral ids to not replace
-    # each others' connections.
-    omq.ephemeral_routing_id = True
+        self.manager = Manager()
+        self.test_suite = False
 
-    return omq
+        if uwsgi.mule_id() != 0:
+            uwsgi.opt['mule'].setup_omq(self)
+            return
 
+        from .web import app  # Imported here to avoid circular import
 
-# Postfork for workers: we start oxenmq and connect to the mule process
-@postfork
-def start_oxenmq():
-    try:
-        import uwsgi
-    except ModuleNotFoundError:
-        return
+        app.logger.debug(f"Starting oxenmq connection to mule in worker {uwsgi.worker_id()}")
 
-    global omq, mule_conn
+        self._omq.start()
+        app.logger.debug("Started, connecting to mule")
+        self.mule_conn = self._omq.connect_remote(oxenmq.Address(config.OMQ_INTERNAL))
 
-    omq = make_omq()
+        app.logger.debug(f"worker {uwsgi.worker_id()} connected to mule OMQ")
 
-    if uwsgi.mule_id() != 0:
-        from . import mule
+        global omq_global
+        omq_global = self
 
-        mule.setup_omq()
-        return
+    def add_bot(self):
+        self.manager.add_bot()
+        # TODO: add omq logic
 
-    from .web import app  # Imported here to avoid circular import
+    def remove_bot(self):
+        self.manager.remove_bot()
+        # TODO: add omq logic
 
-    app.logger.debug(f"Starting oxenmq connection to mule in worker {uwsgi.worker_id()}")
+    def send_mule(self, command, *args, prefix="worker."):
+        """
+        Sends a command to the mule from a worker (or possibly from the mule itself).  The command will
+        be prefixed with "worker." (unless overridden).
 
-    omq.start()
-    app.logger.debug("Started, connecting to mule")
-    mule_conn = omq.connect_remote(oxenmq.Address(config.OMQ_INTERNAL))
+        Any args will be bt-serialized and send as message parts.
+        """
+        
+        if prefix:
+            command = prefix + command
```

Comment:
i feel like `prefix` should not require a trailing `.` to prevent footcannons.
then we can have `command = f"{prefix}.{command}"` here

Reply:
agreed. fixed